src/System.Management.Automation/utils/PsUtils.cs

```diff
@@ -90,45 +90,89 @@ internal static ProcessModule GetMainModule(Process targetProcess)
-        internal static Process GetParentProcess(Process current)
-        {
+        /// <param name="processId">the id of the process whose parent id should be retreived</param>
+        /// <returns>0 on failure, else a parent process id</returns>
+        internal static int GetParentProcessId(int processId)
+        {            
             int parentPid = 0;
-
 #if !UNIX
-            PlatformInvokes.PROCESSENTRY32 pe32 = new PlatformInvokes.PROCESSENTRY32 { };
-            pe32.dwSize = (uint)ClrFacade.SizeOf<PlatformInvokes.PROCESSENTRY32>();
+            // This is a common query (parent id for the current process)
+            // Use cached value if available
+            if (processId == CurrentProcessId && CurrentParentProcessId.HasValue)
+            {
+                return CurrentParentProcessId.Value;
+            }
+
+            parentPid = GetParentProcessIdNative(processId);
+
+            // cache the current process parent pid if it hasn't been done yet
+            if (processId == CurrentProcessId && !CurrentParentProcessId.HasValue)
+            {
+                CurrentParentProcessId = parentPid;
+            }
+#endif
+            return parentPid;
+        }
+
+        /// <summary>
+        /// Helper function to get the parent process id from the win32 Toolhelp apis
+        /// </summary>
+        /// <remarks>
+        /// This could potentially be replaced with calls to NTQueryInformationProcess
+        /// </remarks>
+        /// <param name="processId">The child process id to get the parent from</param>
+        /// <returns></returns>
+        private static int GetParentProcessIdNative(int processId)
+        {
+            int parentPid = 0;
+            PlatformInvokes.PROCESSENTRY32 pe32 = new PlatformInvokes.PROCESSENTRY32 {};
+            pe32.dwSize = (uint) ClrFacade.SizeOf<PlatformInvokes.PROCESSENTRY32>();
 
-            using (PlatformInvokes.SafeSnapshotHandle hSnapshot = PlatformInvokes.CreateToolhelp32Snapshot(PlatformInvokes.SnapshotFlags.Process, (uint)current.Id))
+            using (
+                PlatformInvokes.SafeSnapshotHandle hSnapshot =
+                    PlatformInvokes.CreateToolhelp32Snapshot(PlatformInvokes.SnapshotFlags.Process, (uint) processId))
             {
                 if (!PlatformInvokes.Process32First(hSnapshot, ref pe32))
                 {
                     int errno = Marshal.GetLastWin32Error();
                     if (errno == PlatformInvokes.ERROR_NO_MORE_FILES)
                     {
-                        return null;
+                        return parentPid;
                     }
                 }
                 do
                 {
-                    if (pe32.th32ProcessID == (uint)current.Id)
+                    if (pe32.th32ProcessID == (uint) processId)
                     {
-                        parentPid = (int)pe32.th32ParentProcessID;
+                        parentPid = (int) pe32.th32ParentProcessID;
```

Comment:
PG do not accept formatting without code changes. It would be good to revert this.